File: 26-OutaTimeRedux/OutaTimeRedux/TimeCircuitsViewController.m
//
//  ViewController.m
//  OutaTimeRedux
//
//  Created by Chris Stomp on 11/9/15.
//  Copyright © 2015 The Iron Yard. All rights reserved.
//

#import "TimeCircuitsViewController.h"
#import "DatePickerViewController.h"

@interface TimeCircuitsViewController ()

@property (weak, nonatomic) IBOutlet UILabel *destinationTimeLabel;
@property (weak, nonatomic) IBOutlet UILabel *presentTimeLabel;
@property (weak, nonatomic) IBOutlet UILabel *lastTimeDepartedLabel;
@property (weak, nonatomic) IBOutlet UILabel *speedLabel;
@property (nonatomic) int currentSpeed;
@property (nonatomic) NSTimer *timer;


-(IBAction)setDestinationTimeTapped:(UIButton *)sender;
-(IBAction)travelBackTapped:(UIButton *)sender;

NSString *dateWasChosen(NSDate *);

@end

@implementation TimeCircuitsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.currentSpeed = 0;
    self.presentTimeLabel.text = [self dateFormat:[NSDate date]];
    self.speedLabel.text = @"O MPH";
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    DatePickerViewController* datePickerVC = (DatePickerViewController *)segue.destinationViewController;
    datePickerVC.delegate = self;
}

- (void) startTimer
    {
        if (self.timer == nil)
        {
            self.timer = [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(updateSpeed) userInfo:nil repeats:YES];
        }
    }

- (void) updateSpeed
{
    if (self.currentSpeed < 88)
    {
        self.currentSpeed += 1;
    }
    else
    {
        [self.timer invalidate];
        self.currentSpeed = 0;
        self.lastTimeDepartedLabel.text = self.presentTimeLabel.text;
        //self.presentTimeLabel.text;
        self.presentTimeLabel.text = self.destinationTimeLabel.text;
    }
    
    self.speedLabel.text =[NSString stringWithFormat:@"%d MPH", self.currentSpeed];
}


#pragma mark - Action Handlers


- (IBAction)travelBackTapped:(UIButton *)sender
{
    [self startTimer];
}

#pragma mark - Date Picker Delegate

- (void) dateWasChosen:(NSDate *)date;
{
    NSString *dateString = [self dateFormat:date];
    self.destinationTimeLabel.text = dateString;
}

- (NSString *) dateFormat:(NSDate *)date;
{
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc]init];
    
    [dateFormatter setDateFormat:@"MMM dd YYYY"];
    NSString *formattedTime = [dateFormatter stringFromDate:date];
    NSString *capitalizedTime = [formattedTime uppercaseString];
    return capitalizedTime;
}

@end

File: 04-OutaTime/04-OutaTime/TimeCircuitsViewController.swift
//
//  TimeCircuitsViewController.swift
//  04-OutaTime
//
//  Created by Chris Stomp on 11/5/15.
//  Copyright © 2015 The Iron Yard. All rights reserved.
//

import UIKit

@objc protocol DatePickerDelegate
{
    func dateWasChosen(date: NSDate)
}

class TimeCircuitsViewController: UIViewController, DatePickerDelegate
{
    
    @IBOutlet weak var destinationTimeLabel: UILabel!
    @IBOutlet weak var presentTimeLabel: UILabel!
    @IBOutlet weak var lastTimeDepartedLabel: UILabel!
    @IBOutlet weak var speedLabel: UILabel!
    
    var timer: NSTimer?
    var currentSpeed = 0
    let dateFormatter = NSDateFormatter()
    
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        title = "Time Circuits"
        
        let dateString = formatTime(NSDate())
        
        destinationTimeLabel.text = "_ _ _ _ _ _ _ _ _"
        presentTimeLabel.text = dateString
        speedLabel.text = "\(currentSpeed) MPH"
        lastTimeDepartedLabel.text = "_ _ _ _ _ _ _ _ _"

    }


    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?)
    {
        if segue.identifier == "ShowDatePickerSegue"
        {
            let datePickerVC = segue.destinationViewController as! DatePickerViewController
            datePickerVC.delegate = self
        }
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */
    
    @IBAction func travelBackTapped(sender: UIButton)
    {
        
        startTimer()
        updateSpeed()
    }
    
    func formatTime(timeToFormat: NSDate) -> String
    {
        self.dateFormatter.dateStyle = .MediumStyle
        dateFormatter.dateFormat = "MMM dd, yyyy"
        let formattedTime = dateFormatter.stringFromDate(timeToFormat)
        return String(formattedTime)
    }
    
    func dateWasChosen(date: NSDate)
    {
        destinationTimeLabel.text = dateFormatter.stringFromDate(date)
    }
    
    func startTimer()
    {
        if timer == nil
        {
           timer = NSTimer.scheduledTimerWithTimeInterval(0.1, target: self, selector: "updateSpeed", userInfo: nil, repeats: true)
        }
        
    }
    
    func updateSpeed()
    {
        if currentSpeed < 88
        {
            currentSpeed += 1
            speedLabel.text = String(currentSpeed)
        }
        else
        {
            stopTimer()
        }
    }
    
    func stopTimer()
    {
        timer?.invalidate()
        speedLabel.text = "0" //String(currentSpeed)
        presentTimeLabel.text = destinationTimeLabel.text
        lastTimeDepartedLabel.text = formatTime(NSDate())
        
    }
    
}
